carstore/pgbackend.go

```diff
@@ -0,0 +1,376 @@
+		if err := writeShardBlocks(ctx, &sh, w); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (pgb *PgBackend) PutShard(ctx context.Context, shard *CarShard, brefs []map[string]any, rmcids map[cid.Cid]bool) error {
+	// TODO: there should be a way to create the shard and block_refs that
+	// reference it in the same query, would save a lot of time
+	tx := pgb.db.WithContext(ctx).Begin()
+
+	if err := tx.WithContext(ctx).Create(shard).Error; err != nil {
+		return fmt.Errorf("failed to create shard in DB tx: %w", err)
+	}
+
+	for _, ref := range brefs {
+		ref["shard"] = shard.ID
+	}
+
+	if err := createBlockRefs(ctx, tx, brefs); err != nil {
+		return fmt.Errorf("failed to create block refs: %w", err)
+	}
+
+	if len(rmcids) > 0 {
+		cids := make([]cid.Cid, 0, len(rmcids))
+		for c := range rmcids {
+			cids = append(cids, c)
+		}
+
+		if err := tx.Create(&staleRef{
+			Cids: packCids(cids),
+			Usr:  shard.Usr,
+		}).Error; err != nil {
+			return err
+		}
+	}
+
+	err := tx.WithContext(ctx).Commit().Error
+	if err != nil {
+		return fmt.Errorf("failed to commit shard DB transaction: %w", err)
+	}
+
+	return nil
+}
+
+func createBlockRefs(ctx context.Context, tx *gorm.DB, brefs []map[string]any) error {
+	ctx, span := otel.Tracer("carstore").Start(ctx, "createBlockRefs")
+	defer span.End()
+
+	if err := createInBatches(ctx, tx, brefs, 2000); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func generateInsertQuery(data []map[string]any) (string, []any) {
+	placeholders := strings.Repeat("(?, ?, ?),", len(data))
+	placeholders = placeholders[:len(placeholders)-1] // trim trailing comma
+
+	query := "INSERT INTO block_refs (\"cid\", \"offset\", \"shard\") VALUES " + placeholders
+
+	values := make([]any, 0, 3*len(data))
+	for _, entry := range data {
+		values = append(values, entry["cid"], entry["offset"], entry["shard"])
+	}
+
+	return query, values
+}
+
+// Function to create in batches
+func createInBatches(ctx context.Context, tx *gorm.DB, data []map[string]any, batchSize int) error {
+	for i := 0; i < len(data); i += batchSize {
+		batch := data[i:]
+		if len(batch) > batchSize {
+			batch = batch[:batchSize]
+		}
+
+		query, values := generateInsertQuery(batch)
+
+		if err := tx.WithContext(ctx).Exec(query, values...).Error; err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (pgb *PgBackend) GetAllShards(ctx context.Context, usr models.Uid) ([]CarShard, error) {
+	var shards []CarShard
+	if err := pgb.db.Order("seq asc").Find(&shards, "usr = ?", usr).Error; err != nil {
+		return nil, err
+	}
+	return shards, nil
+}
+
+func (pgb *PgBackend) WipeUserData(ctx context.Context, user models.Uid) error {
+	var shards []*CarShard
+	if err := pgb.db.Find(&shards, "usr = ?", user).Error; err != nil {
+		return err
+	}
+
+	if err := pgb.deleteShards(ctx, shards); err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (pgb *PgBackend) deleteShards(ctx context.Context, shs []*CarShard) error {
```

Comment:
intentionally non-public (not capitalized?)

Reply:
oh, no. fixing